planner/core/task.go

```diff
@@ -200,18 +346,51 @@ func finishCopTask(ctx sessionctx.Context, task task) task {
 	if !ok {
 		return task
 	}
-	// FIXME: When it is a double reading. The cost should be more expensive. The right cost should add the
-	// `NetWorkStartCost` * (totalCount / perCountIndexRead)
+	// copTasks are run in parallel, to make the estimated cost closer to execution time, we amortize
+	// the cost to cop iterator workers. According to `CopClient::Send`, the concurrency
+	// is Min(DistSQLScanConcurrency, numRegionsInvolvedInScan), since we cannot infer
+	// the number of regions involved, we simply use DistSQLScanConcurrency.
+	copIterWorkers := float64(t.plan().context().GetSessionVars().DistSQLScanConcurrency)
 	t.finishIndexPlan()
+	// Network cost of transferring rows of table scan to TiDB.
 	if t.tablePlan != nil {
-		t.cst += t.count() * netWorkFactor
+		t.cst += t.count() * netWorkFactor * t.tableStats.HistColl.GetAvgRowSize(t.tablePlan.Schema().Columns, false)
 	}
+	t.cst /= copIterWorkers
 	newTask := &rootTask{
 		cst: t.cst,
 	}
 	if t.indexPlan != nil && t.tablePlan != nil {
 		p := PhysicalIndexLookUpReader{tablePlan: t.tablePlan, indexPlan: t.indexPlan}.Init(ctx)
 		p.stats = t.tablePlan.statsInfo()
+		// Add cost of building table reader executors. Handles are extracted in batch style,
+		// each handle is a range, the CPU cost of building copTasks should be:
+		// (indexRows / batchSize) * batchSize * cpuFactor
+		// Since we don't know the number of copTasks built, ignore these network cost now.
+		indexRows := t.indexPlan.statsInfo().RowCount
+		newTask.cst += indexRows * cpuFactor
+		// Add cost of worker goroutines in index lookup.
+		numTblWorkers := float64(t.indexPlan.context().GetSessionVars().IndexLookupConcurrency)
+		newTask.cst += (numTblWorkers + 1) * concurrencyFactor
+		// When building table reader executor for each batch, we would sort the handles. CPU
+		// cost of sort is:
+		// cpuFactor * batchSize * Log2(batchSize) * (indexRows / batchSize)
+		indexLookupSize := float64(t.indexPlan.context().GetSessionVars().IndexLookupSize)
+		batchSize := math.Min(indexLookupSize, indexRows)
+		if batchSize > 2 {
+			sortCPUCost := (indexRows * math.Log2(batchSize) * cpuFactor) / numTblWorkers
```

Comment:
Is it possible that `batchSize` is 0?

Reply:
The check condition `batchSize > 2` can protect us?